pkg/distributor/otlp/cmd/generate/main.go

```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package main
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/alecthomas/kingpin/v2"
+
+	// Dummy dependency to ensure these files get vendored and we can use them as transformation sources.
+	_ "github.com/prometheus/prometheus/storage/remote/otlptranslator/prometheusremotewrite"
+)
+
+func main() {
+	app := kingpin.New("generate", "Generate OTLP translation code.")
+
+	generate := func(*kingpin.ParseContext) error {
+		sources, err := copySources()
+		if err != nil {
+			return err
+		}
+		if err := patchSources(sources); err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	app = app.Action(generate)
+	kingpin.MustParse(app.Parse(os.Args[1:]))
+}
+
+func patchSources(sources []string) error {
+	if out, err := exec.Command("gopatch", "-p", "cmd/generate/mimirpb.patch", "./").CombinedOutput(); err != nil {
+		return fmt.Errorf("failed to execute gopatch: %s", out)
+	}
+	sed := "sed"
+	if runtime.GOOS == "darwin" {
+		sed = "gsed"
+	}
```

Comment:
@pstibrany how would you do this check in shell?

Reply:
I recall it as being trickier than you'd think.

Reply:
We use this in our scripts: https://github.com/grafana/mimir/blob/50fdbdf2bc5d56739a5a71825423fbc18ae97a13/tools/release/common.sh#L9-L10